pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -693,6 +693,11 @@ public void connectionFailed(PulsarClientException exception) {
 
         CompletableFuture<Void> closeFuture = new CompletableFuture<>();
         ClientCnx cnx = cnx();
+        if (null == cnx) {
```

Comment:
There are few cleanup operations to complete when the "close" is done, perhaps the part at line  704-713 can be extracted into a method.

Also, `cmd` should be created after this check, to avoid leaking it.

Reply:
@merlimat I addressed your comment. can you take a look again?